### ISSUE TITLE
Multiple uploads caused some issues

### ISSUE THREAD

**Issue by slifty** [PermanentOrg/sftp-service#99]
The root cause is not clear, but during a demo this afternoon an attempt to upload 6 `jpeg` files resulted in some errors.

This could be related to caching, it could be related to download name changes, or something else.

This issue is just a placeholder for me to dig in and learn more!

**Comment by slifty** [comment]
I looked at the server logs and it isn't immediately clear what the problem is (the prod server has a log level of info so not all details are there).

That said, I'm seeing deletion requests which indicates to me that rclone thinks the files failed to upload properly.  I suspect this is related to #97 and might suggest that we need to do even more / need to somehow make sure that at *least* the size of a file is populated properly on permanent before we consider a file uploaded.

Other things to consider here is the what kinds of issues might be caused here by caching.

**Comment by slifty** [comment]
I've spent some time debugging this situation.

Some notes:

1. It happens unpredictably -- and the more files being uploaded, the more likely it will happen.
2. When a file hangs it seems to indefinitely (I waited around 10 minutes before giving up).
3. The hang consistently occurs at the point of a `node-fetch` call, however the specific endpoint being accessed / the specific call does not seem to matter.  This is to say: the hang might happen when looking up a folder, looking up an archive record, creating an S3 URL.  The only consistent behavior I've seen is that the calls are to permanent resources (as opposed to the POST to amazon S3).  That being said this could be due to the fact that calls to permanent make up the vast majority (e.g. 99%) of the fetch calls associated with SFTP.

All this is to say: The root cause from the SFTP service's perspective is that certain API calls to the permanent backend are not resolving -- that is, they are neither succeeding or failing, but simply sitting open without timing out.

The unpredictable nature of the hangs indicates to me that the issue is not with any particular call, but rather something lower level (e.g. HTTP or database connection pools, locked resources, etc.)

I currently have two "obvious" paths of exploration:

1. *Looking into the HTTP library that we are using* (`node-fetch`).  This library could possibly have a bug in it that results in connections to silently fail in certain circumstances.  This is not compelling because (A) the library itself is already a major version ahead of what we are using, though we cannot upgrade due to node version, and (B) we are planning to get rid of the library anyway. I would prefer to use this as an opportunity to update the version of NodeJS to 18.x, as Node recently added native support for fetch and we could then remove the dependency.

2. *Looking into the permanent backend more closely*.  I should note that there are no obvious backend errors in the existing sentry or log files.

My hope is that by using a native fetch library either I will (A) fix the issue, or (B) make it easier to debug the permanent back end.  The current library has some limitations related to errors in the event of timeouts.

**Comment by slifty** [comment]
This was *incredibly* intense to debug.  It was so very very unpredictable.

This is breaking news, but it's the end of the day and so I want to document it even though I have not yet done a full verification that everything is OK (e.g. removed all debug code and then implemented the identified fix).

There are a few issues going on.

1) If anything goes wrong during upload, rclone will try to delete the file -- our delete handlers do not respond at all, and so rclone sits and waits.  This is a bug (and may actually have been the root cause of the bug that caused us to open this issue), but the thing that was driving me in circles, I believe, is:

2) This line: https://github.com/PermanentOrg/sftp-service/blob/5d7981a6c1700f14af0c1069c8a4a350af86240b/src/classes/SftpSessionHandler.ts#L332

If you try to `fs.close` or `fs.closeSync` a file descriptor that is in use ANYWHERE ELSE then it results in "`unknown behavior`" according to the node documentation.

In our case, I believe the "unknown behavior" was *random IO will silently fail*.  I was seeing requests never arrive to our server, I was seeing requests to Amazon S3 never resolve, I was seeing server responses never get consumed by our processes.  (I even think I saw a `setTimeout` fail to resolve, but I'm going to just smile and nod at that one).

Bottom line was that the only pattern seemed to be that I/O was occasionally failing.

Commenting out that close resolved those problems.  I do need to make sure that temp files are safely being opened and closed, but clearly this isn't the way.  I'm going to make a separate issue for this, since I clearly need to take a more comprehensive look at the use and management of temporary files.
